### PR TITLE
Add improved tracking to the marketplace

### PR DESCRIPTION
**Pull request by kdevnel**
### Changes proposed in this Pull Request:

Closes https://github.com/Automattic/woocommerce.com/issues/18294

The issue linked provides improvements to the interaction that are tracked in the marketplace page. The initial priority is to ensure that all the tracking that currently exists in the PHP marketplace, is still present in the React marketplace. This PR handles that and adds some improvement to that tracking to better suit the new marketplace structure.

### How to test the changes in this Pull Request:

1. Checkout this branch and build the project
2. Visit any WC Admin page such as `/wp-admin/admin.php?page=wc-admin` ([wp-env link](http://localhost:8888/wp-admin/admin.php?page=wc-admin))
3. Open the browser console, the paste and run the following: `localStorage.setItem( 'debug', 'wc-admin:*' );` to allow tracks events from React to show in the console.
4. Visit the marketplace pages and see the various events are recording correctly to the console. They should make sense based on the tables below. Note: the legacy events are mapped to the logic of the events listed [here](https://github.com/woocommerce/woocommerce/blob/9a0e91477eadc4b4a6027197edb4a5ce1d48548c/plugins/woocommerce/includes/tracks/events/class-wc-extensions-tracking.php#L31-L48).
5. Particularly, check for doubling of event triggering that may skew data.
6. Test around the issue

**Legacy events structure**
| Event | Name | Properties |
|--------|--------|--------|
| Discover tab | `extensions_view` | `section: '_featured'` `version: '2'`|
| Browse (extensions) tab | `extensions_view` |  `section: '_all'` `version: '2'` |
| Browse tab with category selected | `extensions_view` |  `section: category name` `version: '2'` |
| Themes tab | `extensions_view` |  `section: 'themes'` `version: '2'` |
| Themes tab with category selected | no event since this isn't currently supported | N/A |
| Search results | `extensions_view_search` |  `section: 'search'` `search_term: search query entered` `version: "2"` |
| Search results > View All button | no event since this isn't currently supported | N/A |
| Search results > View All button with category | no event since this isn't currently supported | N/A |
| Subscriptions tab | `subscriptions_view` |  `section: 'helper'` `version: '2'` |

**New events structure**
Note, all events have the same name – `marketplace_v2_view` so the name column isn't shown.
| Event | Properties |
|--------|--------|
| Discover tab | `view: 'discover'` |
| Browse (extensions) tab | `view: 'extensions'` |
| Browse tab with category selected | `view: extensions` `category: category name` |
| Themes tab | `view: 'themes'` |
| Themes tab with category selected | `view: 'themes'` `category: category name` |
| Search results | `view: 'search'` `searchTerm: search query entered` |
| Search results > View All button | `view: 'search'` `searchTerm: search query entered` `productType: 'extension' or 'theme'` |
| Search results > View All button with category | `view: 'search'` `searchTerm: search query entered` `productType: 'extension' or 'theme'` `category: category name` |
| Subscriptions tab | `view: 'my-subscriptions'` |

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [x] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

Ensure new marketplace works with the existing tracking events.

</details>

### CODE REVIEW

File: plugins/woocommerce-admin/client/marketplace/components/content/content.tsx [hunk 1]
```diff
@@ -27,6 +31,18 @@ export default function Content(): JSX.Element {
 	// Get the content for this screen
 	useEffect( () => {
 		const abortController = new AbortController();
+
+		// we are recording both the new and legacy events here for now
+		// they're separate methods to make it easier to remove the legacy one later
+		const marketplaceViewProps = {
+			view: selectedTab,
+			searchTerm: query?.term,
+			productType: query?.section,
```
Comment (andfinally) [inline plugins/woocommerce-admin/client/marketplace/components/content/content.tsx]: Tracks doesn't like camel case – it compares against [this](https://github.com/woocommerce/woocommerce/blob/f05e8efc17683b56d4941adb702baa16fe8e8cf5/plugins/woocommerce/includes/tracks/class-wc-tracks-event.php#L26) regex and throws error `A valid prop name must be specified`.

```suggestion
			search_term: query?.term,
			product_type: query?.section,
```

File: plugins/woocommerce-admin/client/marketplace/components/products/products.tsx [hunk 2]
```diff
@@ -57,6 +58,11 @@ export default function Products( props: ProductsProps ): JSX.Element {
 	const showAllButton = query.tab === 'search' && ! query.section;
 
 	function showSection( section: ProductType ) {
+		recordMarketplaceView( {
+			view: 'search',
+			searchTerm: query?.term,
+			productType: section,
```
Comment (andfinally) [inline plugins/woocommerce-admin/client/marketplace/components/products/products.tsx]:
```suggestion
			search_term: query?.term,
			product_type: section,
```

File: plugins/woocommerce-admin/client/marketplace/utils/tracking.tsx [hunk 3]
```diff
@@ -0,0 +1,77 @@
+/**
+ * External dependencies
+ */
+import { recordEvent } from '@woocommerce/tracks';
+
+interface MarketplaceViewProps {
+	view: string;
+	searchTerm?: string;
+	productType?: string;
```
Comment (andfinally) [inline plugins/woocommerce-admin/client/marketplace/utils/tracking.tsx]:
```suggestion
	search_term?: string;
	product_type?: string;
```

**Review by andfinally — CHANGES_REQUESTED** [review]
We need to change this PR to track trunk – we won't be merging the subscriptions feature branch yet!

File: plugins/woocommerce-admin/client/marketplace/utils/tracking.tsx [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+/**
+ * Record a marketplace view event.
+ * This is a new event that is easier to understand and implement consistently
+ */
+function recordMarketplaceView( props: MarketplaceViewProps ) {
+	const eventProps = { ...props };
+
+	recordEvent( 'marketplace_v2_view', eventProps );
```
Comment (andfinally) [inline plugins/woocommerce-admin/client/marketplace/utils/tracking.tsx]: I would be comfortable with calling this `marketplace_view`, as it won't be too hard to find out when we introduced it.

**Review by andfinally — COMMENTED** [review]
I've finally figured out how tracks events are sent - `https://stats.wp.com/w.js` makes a request like `https://pixel.wp.com/t.gif?...` 

When the properties aren't present they are showing up as `'undefined'` in the pixel request – I wonder if we should default them to empty string? For example, on http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=extensions&path=%2Fextensions.

image

To view this, I commented out this bit in [packages/js/tracks/src/index.ts](https://github.com/woocommerce/woocommerce/blob/30151a6cf789c8276033e900973837598b660450/packages/js/tracks/src/index.ts#L52-L55):

```js
	if ( isDevelopmentMode ) {
		// window.wcTracks.validateEvent( eventName, eventProperties );
		// return false;
	}
```

File: plugins/woocommerce-admin/client/marketplace/components/content/content.tsx (hunk 1 above)
Comment (kdevnel) [reply plugins/woocommerce-admin/client/marketplace/components/content/content.tsx]: Ah, good point! I'll rework things to fit snake case then.

**Comment by kdevnel** [comment]
> We need to change this PR to track trunk – we won't be merging the subscriptions feature branch yet!

Yeah, that's true @andfinally. I'll need to rebase this, too, so we don't incidentally bring in parts of the feature branch when we merge! I'll work on that. 

> When the properties aren't present they are showing up as 'undefined' in the pixel request – I wonder if we should default them to empty string? For example, on http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=extensions&path=%2Fextensions.

Regarding this - you raise a good point. I was actually looking for a concise way to not set these properties at all when they're undefined, but I couldn't for the life of me figure out a way to do it. If you have any thoughts, I'm all ears. If not, I will set these as empty strings by default instead.

**Review by mlaetitia — CHANGES_REQUESTED** [review]
Same issue as Gareth - this is what I see in my console:

Xnip2023-10-24_10-27-57

**Comment by kdevnel** [comment]
> Same issue as Gareth - this is what I see in my console:
> 
> Xnip2023-10-24_10-27-57

This is what happens when you filter errors in your console while working on stuff and forget to turn them back on 😵‍💫

**Comment by andfinally** [comment]
> I was actually looking for a concise way to not set these properties at all when they're undefined, but I couldn't for the life of me figure out a way to do it. 

This in `content.tsx` does the trick for me:

```js
		const marketplaceViewProps = {
			view: selectedTab || '',
			search_term: query?.term || '',
			product_type: query?.section || '',
			category: query?.category || '',
		};
```

image

File: plugins/woocommerce-admin/client/marketplace/utils/tracking.tsx [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+	if ( props.view === 'discover' ) {
+		oldEventName = 'extensions_view';
+		oldEventProps.section = '_featured';
+	}
+
+	if ( props.view === 'extensions' ) {
+		oldEventName = 'extensions_view';
+
+		oldEventProps.section = props?.category;
```
Comment (andfinally) [inline plugins/woocommerce-admin/client/marketplace/utils/tracking.tsx]:
```suggestion
		oldEventName = 'extensions_view';
		oldEventProps.section = props?.category;
```

**Comment by kdevnel** [comment]
Okay, I addressed all PR feedback and I've rebased on `trunk` now. I am working on additional tracking but I'll likely do those as a separate PR so we can focus on getting the core tracking in place.

**Review by andfinally — COMMENTED** [review]
Thanks for the tweaks, looking nice!

If it's easy, it would be neat to omit a property if it has an empty value. Sorry if I helped create this situation with my earlier suggestion!

I also noticed that viewing say http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=extensions&path=%2Fextensions clocks up two pairs of events. Maybe when all properties are empty we can assume the event is just triggered by the initial mount, and skip it?

image

Also... events are showing up in Tracks live view – yay!

**Comment by kdevnel** [comment]
@andfinally I've made all your suggested changes. I also happened to find a bug that was causing “View All” to trigger 2 events for the new event type and also trigger an old event with the same name as a normal search. I stopped both of those, so we don't artificially inflate event counts.

Also found a couple of nice ways to handle conditionally assigning properties.

**Review by mlaetitia — COMMENTED** [review]
On my end it seems that all the legacy events are getting properly recorded but the new ones aren't.
It seems that only category filters are getting recorded on the new event - is this intended?

**Review by andfinally — COMMENTED** [review]
I pushed a few small changes in 1f67eb8d5c20748655a0173b3a3aca1c45163f96 which I hope will fix the problem of multiple page view events without breaking anything else! 🙏 The `useEffect` hook we're using to track page views was running twice when the URL changed and the `selectedTab` state changed in response. So I've changed the hook to depend on the tab in the query string instead.

We still have "View All" bug in search results you mentioned – maybe I've reintroduced it, heh. The debugger suggests this is happening because after the first render, we change the `section` query string param. This is the hook's fifth dependency. But we can worry about that another time.

image

I'll give this another go tomorrow!

**Review by andfinally — APPROVED** [review]
I pushed a few more small changes in [`6802488` (#40951)](https://github.com/woocommerce/woocommerce/pull/40951/commits/680248836cfa152b321d8d48c61f457e31e14318).

* Removed `recordMarketplaceView` from `showSection` in `Products` component. We already record the view in the `useEffect` hook  in `Content`. This fixes the duplication with that view.
* Only setting default category `_all` for `search` view if `product_type` (i.e. `query.section`) is also set. The previous method was setting the `_all` category on the initial search results, where category doesn't appear.
* Removed `! props.view` from condition in `recordLegacyTabView`. It meant we weren't recording the legacy view for the initial default marketplace page.

Everything seems to be working OK! @mlaetitia, hopefully you'll now see both sets of events.

**Comment by kdevnel** [comment]
Okay, one final change I believe. @mlaetitia you should be able to use the table in the PR description to confirm what events should fire and when. Just to note - run the build and then leave the marketplace page and return. This can sometimes help. I'm not sure if the tracks event logging caches something that causes issues occasionally.

**Review by andfinally — APPROVED** [review]
Thanks for the latest changes! Everything's working nicely.

Pushed a small change to remove an unused import in [`58fe816` (#40951)](https://github.com/woocommerce/woocommerce/pull/40951/commits/58fe816b8fbc4bbce22924eefdee767d198653e2).

Before we merge, it might be nice to squash the commits to one, or do an interactive rebase to squash the small ones.

**Review by Dan-Q — APPROVED** [review]
Works nicely. Love all the comments.

`plugins/woocommerce-admin/client/marketplace/utils/tracking.tsx` doesn't need to be a `.tsx` file. It could just be a `.ts` and therefore use a simpler Babel module to compile it. Not a big deal though!

Love this.

**Review by mlaetitia — APPROVED** [review]
All working now! Thanks !

**Pull request merged**